### PR TITLE
Add collections

### PR DESCRIPTION
**Pull request by asledz**
@psarna @piodul 
Me and @Kejmer are sending what we came up with. It is still not completed - we need to discuss some further details, will do so on the call tomorrow.
In the meantime we are sending this for pre-review.

### CODE REVIEW

File: examples/cql_collections.rs
Comment (psarna) [inline examples/cql_collections.rs:35]: Why do you cast to a hash map of i32 -> String, if the column is specified above as `map<text, text>`? Shouldn't the column instead be defined as `map<int, text>` in the database?

**Comment by psarna** [comment]
Also, github doesn't allow putting comments directly on commit messages, so I'll do it here: this repository is already public, so we need to adhere to commit messages rules (unless a commit is really trivial). Commit messages should:
1. explain what's changed
2. explain why it needed to be changed

Even for work in progress, it makes sense to spend a moment to create comprehensible commit messages in order to make the review process easier. It will have to be done eventually anyway, so why not start early :)

Please take a look at `git log` (or browse https://github.com/scylladb/scylla-rust-driver/commits/) to see how the messages usually look like.

File: scylla/src/frame/response/result.rs
Comment (psarna) [inline scylla/src/frame/response/result.rs:404]: It's not allowed. When in doubt, just run Scylla, send the query and find out :)

**Comment by psarna** [comment]
@asledz the only test I see is `test_cql_panic` - perhaps you haven't pushed all commits yet? We'd like to have test cases for all top-level collection types.

As for the next steps to finish before Sunday:
1. All commit messages should be amended to address this comment: https://github.com/scylladb/scylla-rust-driver/pull/134#issuecomment-749033759 . For editing commit messages, squashing multiple commits into one, etc, it's most convenient to use `git rebase -i`. Here's a tutorial on how to use it to rewrite git history: https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d
2. Collection types should be tested
3. All review requests from today should be addressed as well

Finally, please let us (me and @piodul) know once the changes are pushed by mentioning us with @, since GitHub doesn't always notify about changes happening in pull requests. If you have any issues with rewriting commit messages or anything else, please let us know and we'll help, since we'd very much like to have this pull request delivered this week in order to be ready for crates.io. Good luck! (:

**Comment by asledz** [comment]
@piodul @psarna 
Fixed

File: scylla/src/frame/response/cql_to_rust.rs
Comment (piodul) [inline scylla/src/frame/response/cql_to_rust.rs:92]: That's not what I meant.

We know how many values we are going to insert into `res`. The idea was to set capacity to the number of elements that we will insert. It's not ideal to set the capacity to a constant value, because:

- If we insert less elements, than we will unnecessarily waste space,
- If we insert more, then we will have to reallocate and rehash the hashmap once or more, which is a little costly.

Although this is a performance optimization which does not affect correctness, we care about performance, and the optimization is quite simple to do.

The two lines above should look like this:

```rust
let vec = cql_val.into_pair_vec().ok_or(FromCQLValError::BadCQLType)?;
let mut res = HashMap::with_capacity(vec.len());
```

Comment (Kejmer) [reply scylla/src/frame/response/cql_to_rust.rs]: Done

**Review by psarna — CHANGES_REQUESTED** [review]
Aside from in-code comments:
I see that *all* commits were squashed into one, which is undesirable but acceptable, but the commit message from this single commit is still missing. Please take a look at https://github.com/scylladb/scylla-rust-driver/pull/134#issuecomment-749033759 , where I described how a commit message should look like. In particular, `CQL Types` is not a good title for several reasons - it's too generic and doesn't really say what happened to these types. Consider something like `Implement CQL collection types`. Also, a short description of what exactly was added should end up in the message.

Finally, issue https://github.com/scylladb/scylla-rust-driver/issues/97 was not only about collections, but also about the rest of CQL types, which are not implemented yet - e.g. durations, timestamps, varints, etc. Here's a full list of CQL types: https://docs.scylladb.com/getting-started/types/ . Let's not do it as part of this pull request however, we'll instead assign it to someone as a separate issue during the next sync meeting.

File: scylla/src/frame/response/result.rs
Comment (piodul) [inline scylla/src/frame/response/result.rs:545]: Nit: parentheses around `decoded[0]` are unnecessary. Please remove them here and in other assertions in this file, too.

File: examples/cql_collections.rs
Comment (piodul) [inline examples/cql_collections.rs:39]: The test does not check that serializing maps work - but it should. It should include something like this:

```rust
let teams: HashMap<i32, String> = vec![
    (1, "Ginny".to_owned()),
    (2, "Cho".to_owned()),
].into_iter().collect();

session
    .query(
        "INSERT INTO teams (id, team) VALUES (2, ?)",
        &[teams],
    )
    .await?;
```

There should also be tests like this for lists, sets and tuples.

File: examples/cql_collections.rs
Comment (asledz) [reply examples/cql_collections.rs:44]: Fixed.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs]: Done.

File: examples/cql_collections.rs
Comment (asledz) [reply examples/cql_collections.rs]: Done

File: scylla-macros/src/from_user_type.rs
Comment (asledz) [reply scylla-macros/src/from_user_type.rs:39]: @Kejmer 
I know it was done for some reason, I cannot recall now why  - please explain.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs]: Uncommented.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs]: removed.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs:532]: It got messed up while commiting - it isn't there anymore

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs:575]: Fixed - I messed up the test names. They are correct now.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs:578]: Fixed.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs:545]: Fixed.

File: examples/cql_collections.rs
Comment (asledz) [reply examples/cql_collections.rs:39]: Done.

File: scylla/src/frame/response/result.rs
Comment (asledz) [reply scylla/src/frame/response/result.rs:577]: Fixed.

File: scylla-macros/src/from_user_type.rs
Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: Having HashMap as a key value requires it to has a Hash trait implemented and HashMap doesn't provide default implementation and BTreeMap provides the same functionality and has default implementation for Hash.

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: @psarna

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:39]: @Kejmer where exactly do you need to use HashMap as a key? Can you point me to the line in code? BTreeMap may provide similar functionality, but does not have identical performance characteristics.

**If** we want to migrate from HashMap to BTreeMap, then it cannot be done in the same commit which implements CQL types - it's an unrelated change. But for starters I need to understand why we're interested in storing the whole map as a key - while BTreeMap provides a Hash implementation, it's likely O(n), so it would generally be suboptimal to use it as a key anywhere. Also, we don't need ordering of the fields, so hash map is simply a better choice performance-wise, both in terms of execution time and memory utilization.

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: @psarna 
result.rs:54 has #[derive(Hash)]
cql_to_rust.rs:87 requires first type to be hashable

Since Map accepts any CQLType as key value, it could be a Map as well.

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:39]: Ok, then I'll rather provide a hasher for this hash map (in a separate commit) than change the underlying type. I'm not 100% convinced, since it might be premature optimization, but on the other hand UDTs are themselves an optimization, so we might want to keep their implementation faster if the price is low - implementing a single hasher. But first, I need a second opinion :) @piodul?

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: I'm trying to break this commit into a bunch of smaller ones.

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: @psarna if i remember correctly there is also a problem with implementing Hash trait to HashMap since neither trait or class was declared in our files and one workaround we found was to create a class wrapper that contains only HashMap inside and then implement a Hash function for it, but there is still problem of HashMap being unordered, so  two maps with same keys and same elements can return different Hash values and as a result create future bugs.

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:39]: That's a valid concern (although it's still possible to compute a hash which disregards the order of elements). I'm generally fine with using BTreeMap instead, since there is a sane reasoning behind the change, but please list all these arguments in the commit message of the separate commit which changes the map type.

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:39]: @psarna 
Would it be a problem if I create another PR for this task and close this one? 
I've cherry-picked this one big commit and broke it down into smaller ones, it was easier for me that way.

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:39]: No problem, although you can also simply force-push to this PR's branch, and the git history will get rewritten.

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:39]: (force-push == `git push -f`, or, better yet, `git push --force-with-lease`)

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:28]: @psarna I merged cherry-picked branch with ours strategy. Is everything ok now in your opinion?

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:28]: I have no idea what happened to git history here - we generally want to avoid merges from inside the pull request. The rule of thumb is that if you want to type `git merge` in our repo - don't :) Merging branches is left for the maintainers only, so that we sustain a clear, forward history of changes. Same goes with `git pull`, unless you use `git pull --rebase`.

Please repush without this merge commit. You can do it e.g. like this:
1. Checkout main: `git fetch; git checkout origin/main`
2. Cherry-pick your commits on top of main: `git cherry-pick 351afdf02 14776f4d8a f7e5284144 7247b4e74 c209b2a0c 41c70d79e d30712861a d30712861a`
3. Force-push

Comment (Kejmer) [reply scylla-macros/src/from_user_type.rs:28]: This method doesn't work, I tried to `git push --force-with-lease origin add_collecions` but it doesn't overwrite changes. I think creating new PR would be a less painfull way with clear commit history.

Comment (psarna) [reply scylla-macros/src/from_user_type.rs:28]: fine with me

**Comment by piodul** [comment]
A version of this PR with cleaned up git history was posted [here](https://github.com/scylladb/scylla-rust-driver/pull/141), review should continue there. I'm closing this.

**Pull request closed without merging**
